### PR TITLE
Create a ducks for categories

### LINKED ISSUE #6935: Create a ducks file for categories (reducers, actions)
See also: #4523

---

Similar to #6931, let's create a ducks for the `categories` and add more flow coverage.

**For QA:** please make sure categories load fine, all the category pages but also the list of categories on each landing page and the "hidden" categories pages on Android.

### PR DESCRIPTION
**Pull request by willdurand**
Fixes #6935
~~:warning: depends on #6933~~

---

This PR adds Flow coverage to the `categories` reducer/saga and switch to a ducks file.

### CODE REVIEW

**Comment by willdurand** [comment]
rebased!

**Comment by willdurand** [comment]
rebased!

File: tests/unit/core/reducers/test_categories.js [hunk 1]
```diff
@@ -24,21 +28,19 @@ describe(__filename, () => {
...
-  describe('CATEGORIES_FETCH', () => {
-    it('sets loading', () => {
-      const state = categories(
-        initialState,
-        categoriesFetch({ errorHandlerId: 'some-handler' }),
-      );
-      expect(state.categories).toEqual(null);
-      expect(state.loading).toEqual(true);
-    });
+  it('sets loading', () => {
```
Comment (bobsilverberg) [inline tests/unit/core/reducers/test_categories.js]: Should this be inside the `describe` block for `CATEGORIES_FETCH`?

File: src/core/api/index.js [hunk 2]
```diff
@@ -270,7 +270,9 @@ export function startLoginUrl({
   return `${API_BASE}${apiVersion}/accounts/login/start/${query}`;
 }
 
-export function categories({ api }: {| api: ApiState |}) {
+export type CategoriesParams = {| api: ApiState |};
```
[the remark this answers is not in the document]
Comment (willdurand) [reply src/core/api/index.js]: I have no plan to update this code here, I'll file a new issue for the API code.

File: src/core/reducers/categories.js [hunk 3]
```diff
@@ -1,42 +1,123 @@
...
+}: CategoriesLoadParams): CategoriesLoadAction {
+  return {
+    type: CATEGORIES_LOAD,
+    payload: { results },
+  };
+}
+
+type EmptyCategoryListType = {|
+  [appName: string]: {|
+    [addonType: string]: Array<ExternalCategory>,
```
[the remark this answers is not in the document]
Comment (willdurand) [reply src/core/reducers/categories.js]: `EmptyCategoryListType` is used in the reducer, to create the final state. I've only added types..

File: tests/unit/core/reducers/test_categories.js (hunk 1 above)
Comment (willdurand) [reply tests/unit/core/reducers/test_categories.js]: Yep

File: src/core/reducers/categories.js [hunk 4]
```diff
@@ -1,42 +1,123 @@
...
+type CategoriesFetchParams = {|
+  errorHandlerId: string,
+|};
+
+export type CategoriesFetchAction = {|
+  type: typeof CATEGORIES_FETCH,
+  payload: CategoriesFetchParams,
+|};
+
+export function categoriesFetch({
```
[the remark this answers is not in the document]
Comment (willdurand) [reply src/core/reducers/categories.js]: Yep

File: src/core/reducers/categories.js (hunk 3 above)
[the remark this answers is not in the document]
Comment (willdurand) [reply src/core/reducers/categories.js]: ok I think I understand your comment now.. I changed the main type to `CategoryMapType`.

File: src/core/api/index.js (hunk 2 above)
[the remark this answers is not in the document]
Comment (willdurand) [reply src/core/api/index.js]: See: https://github.com/mozilla/addons-frontend/issues/6978

File: src/core/reducers/categories.js [hunk 5]
```diff
@@ -1,42 +1,123 @@
...
+  weight: number,
+  description: string | null,
+|};
+
+// The absence of strict types is wanted because the logic in the reducer is a
+// mess...
+type CategoryListType = {
+  [appName: string]: {
+    [addonType: string]: {
+      [addonSlug: string]: ExternalCategory,
```
[the remark this answers is not in the document]
Comment (willdurand) [reply src/core/reducers/categories.js]: Yep. The only thing that comes to my mind is: "Rome wasn't built in a day" 😋

File: src/core/api/index.js (hunk 2 above)
[the remark this answers is not in the document]
Comment (bobsilverberg) [reply src/core/api/index.js]: You didn't mention my 2nd comment above in the issue you opened. Do you think it's worth also changing the name when #6978 is addressed?

File: src/core/reducers/categories.js [hunk 6]
```diff
@@ -91,24 +91,24 @@ export function createEmptyCategoryList(): EmptyCategoryListType {
       ...object,
       [appName]: validAddonTypes.reduce((appObject, addonType) => {
         // This is where we define an `EmptyCategoryListType` and not a
-        // `CategoryListType`, because `[addonType]` is initialized with an
+        // `CategoryMapType`, because `[addonType]` is initialized with an
```
Comment (bobsilverberg) [inline src/core/reducers/categories.js]: Nit: I'm not sure we even need this comment anymore, now that they have different names. I think the code might be understandable enough without it, but I'll leave that up to you.

File: src/core/reducers/categories.js [hunk 7]
```diff
@@ -1,42 +1,123 @@
...
+}: FetchCategoriesParams): FetchCategoriesAction {
+  invariant(errorHandlerId, 'errorHandlerId is required');
+
+  return {
+    type: FETCH_CATEGORIES,
+    payload: { errorHandlerId },
+  };
+}
+
+type CategoriesLoadParams = {|
```
Comment (bobsilverberg) [inline src/core/reducers/categories.js]: You missed renaming this one to `LoadCategoriesParams`.

File: src/core/reducers/categories.js [hunk 8]
```diff
@@ -1,42 +1,123 @@
...
+  results: Array<ExternalCategory>,
+|};
+
+type CategoriesLoadAction = {|
```
Comment (bobsilverberg) [inline src/core/reducers/categories.js]: You missed renaming this one to `LoadCategoriesAction`.

File: src/core/api/index.js (hunk 2 above)
[the remark this answers is not in the document]
Comment (willdurand) [reply src/core/api/index.js]: Yep we should do that! I wrote "change the name of the API functions" for that

[the remark this answers is not in the document]
Comment (bobsilverberg) [reply src/core/api/index.js]: Oops, sorry I missed that you said that.

[the remark this answers is not in the document]
Comment (willdurand) [reply src/core/api/index.js]: no worries, it could totally be me not writing proper English :p

File: src/core/reducers/categories.js (hunk 6 above)
Comment (willdurand) [reply src/core/reducers/categories.js]: that's correct

File: src/core/reducers/categories.js (hunk 7 above)
Comment (willdurand) [reply src/core/reducers/categories.js]: oh :(

**Pull request merged**